llm-finetuning/huggingface/hf_deployment_service.py

```diff
@@ -0,0 +1,209 @@
+
+    @property
+    def hf_endpoint(self) -> InferenceEndpoint:
+        """Get the deployed Huggingface inference endpoint.
+
+        Returns:
+            Huggingface inference endpoint.
+        """
+        return get_inference_endpoint(
+            name=self.config.endpoint_name,
+            token=self.config.token,
+            namespace=self.config.namespace,
+        )
+
+    @property
+    def prediction_url(self) -> Optional[str]:
+        """The prediction URI exposed by the prediction service.
+
+        Returns:
+            The prediction URI exposed by the prediction service, or None if
+            the service is not yet ready.
+        """
+        if not self.is_running:
+            return None
+        return self.hf_endpoint.url
+
+    @property
+    def inference_client(self) -> InferenceClient:
+        """Get the Huggingface InferenceClient from Inference Endpoint.
+
+        Returns:
+            Huggingface inference client.
+        """
+        return self.hf_endpoint.client
+
+    def provision(self) -> None:
+        """Provision or update remote Huggingface deployment instance.
+
+        This should then match the current configuration.
+        """
+
+        _ = create_inference_endpoint(
+            name=self.config.endpoint_name,
+            repository=self.config.repository,
+            framework=self.config.framework,
+            accelerator=self.config.accelerator,
+            instance_size=self.config.instance_size,
+            instance_type=self.config.instance_type,
+            region=self.config.region,
+            vendor=self.config.vendor,
+            account_id=self.config.account_id,
+            min_replica=self.config.min_replica,
+            max_replica=self.config.max_replica,
+            revision=self.config.revision,
+            task=self.config.task,
+            custom_image=self.config.custom_image,
+            type=self.config.endpoint_type,
+            namespace=self.config.namespace,
+            token=self.config.token,
+        ).wait(timeout=POLLING_TIMEOUT)
+
+        if self.hf_endpoint.url is not None:
+            logger.info(
+                "Huggingface inference endpoint successfully deployed."
+            )
+        else:
+            logger.info(
+                "Failed to start huggingface inference endpoint service..."
+            )
+
+    def check_status(self) -> Tuple[ServiceState, str]:
+        """Check the the current operational state of the HuggingFace deployment.
+
+        Returns:
+            The operational state of the HuggingFace deployment and a message
+            providing additional information about that state (e.g. a
+            description of the error, if one is encountered).
+        """
+        try:
+            _ = self.hf_endpoint.status
+        except (InferenceEndpointError, HfHubHTTPError):
+            return (ServiceState.INACTIVE, "")
+
+        if self.hf_endpoint.status == "running":
```

Comment:
can we use an enum for these status's? Maybe we have one in zenml already

Reply:
Added it in!